crates/db_views/src/custom_emoji_view.rs

```diff
@@ -0,0 +1,153 @@
+use crate::structs::{CustomEmojiView, CustomEmojiViewForm};
+use diesel::{insert_into, result::Error, BelongingToDsl, ExpressionMethods, GroupedBy, QueryDsl};
+use diesel_async::RunQueryDsl;
+use lemmy_db_schema::{
+  newtypes::LocalSiteId,
+  schema::{
+    custom_emoji::{
+      dsl::{custom_emoji, local_site_id},
+      id,
+    },
+    custom_emoji_keyword::{custom_emoji_id, dsl::custom_emoji_keyword},
+  },
+  source::{
+    custom_emoji::{CustomEmoji, CustomEmojiForm},
+    custom_emoji_keyword::{CustomEmojiKeyword, CustomEmojiKeywordForm},
+  },
+  utils::{get_conn, DbPool},
+};
+
+impl CustomEmojiView {
+  pub async fn get(pool: &DbPool, emoji_id: i32) -> Result<Self, Error> {
+    let conn = &mut get_conn(pool).await?;
+    let emoji = custom_emoji
+      .find(emoji_id)
+      .first::<CustomEmoji>(conn)
+      .await?;
+    let keywords = CustomEmojiKeyword::belonging_to(&emoji)
```

Comment:
Please stop coding from scratch, copy-paste from the other examples in this folder. This is nothing like any of the other examples.

Reply:
This is the recommended way to get a one-to-many relationship, according to diesel https://docs.diesel.rs/diesel/associations/index.html. 

I don't see any other examples in the code base that handle the same situation; closest I've seen is when getting a single entity in one method, then getting its children in a different method, i.e. CommunityView::Read, then CommunityModeratorView::for_community (in read_community.rs). I can technically do it the same way, but this would require 2 queries per custom emoji.

Reply:
Doing two queries just to get some emojis is not going to be good for performance, and these should all stay single-query views. You have two options I can think of that would keep this a single query:


1. Make this a `CustomEmojiKeywordView`. Then you query the keyword table, and do an inner join to the custom_emoji table. This seems to me the best option, since later on, there might be so many emojis, that we'd need to add a search for them as people type in `:` in a text field. This is how we do user and community searches in the text fields.
3. Use something [like `string_agg` to concat the keywords.](https://stackoverflow.com/a/43944/1655478) I don't really like this as much, but it works.

Reply:
Two queries vs 1 isn't a great deal of difference. Diesel recommends this pattern as a way to map a one to many relationship. These particular queries only run when getting SiteRes, so pretty much only on first visit/refresh; not a big deal.

Not sure what you mean by number 1, create a new view in the database? I can condense to one query, but I still need to end up doing a bunch of tedious cpu work to create the resulting structs. Do you have a code example? I haven't seen any better way to map one to many in diesel-land. Also, we don't do searches in the backend, its all done in the frontend after initial load.

On Number 2, I'd very much like to avoid that; not semantically correct and can result in some errors if scope ever changes.

The better way to avoid queries would be to implement caching on the server side, since emojis don't really change that often. Just invalidate cache whenever an emoji changes i guess.

Reply:
Minimizing the number of queries by using inner joins, is much preferred over chaining queries. If we did things this way the front page would be running hundreds of queries, rather than just two or three. Waiting for round-trip responses back from the DB, just to get data you already know is connected, is not good design. That and all these views should be single-query, as they should be equivalent to SQL views. 

Strategy 1 would mean that you get rid of `CustomEmojiView`, and replace it with `CustomEmojiKeywordView`. 

The query would then be:

```rs
custom_emoji_keyword
  .inner_join(custom_emoji::table)
  .select((
    keyword, // (or all the columns if you want)
    custom_emoji::all_columns
  ))
```

Check out the examples in the other db_views.

Reply:
Github is being weird, it didn't do that comment above in the review.

Reply:
Still using two queries.